backends/ebpf/ebpfDeparser.cpp

```diff
@@ -0,0 +1,376 @@
+        DeparserPrepareBufferTranslator(deparser), deparser(deparser) {
+    setName("DeparserHdrEmitTranslator");
+}
+
+void DeparserHdrEmitTranslator::processMethod(const P4::ExternMethod *method) {
+    if (method->method->name.name == p4lib.packetOut.emit.name) {
+        auto decl = method->object;
+        if (decl == deparser->packet_out) {
+            auto expr = method->expr->arguments->at(0)->expression;
+            auto exprType = deparser->program->typeMap->getType(expr);
+            auto headerToEmit = exprType->to<IR::Type_Header>();
+            if (headerToEmit == nullptr) {
+                ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                        "Cannot emit a non-header type %1%", expr);
+            }
+
+            cstring msgStr;
+            builder->emitIndent();
+            builder->append("if (");
+            this->visit(expr);
+            builder->append(".ebpf_valid) ");
+            builder->blockStart();
+            auto program = deparser->program;
+            unsigned width = headerToEmit->width_bits();
+            msgStr = Util::printf_format("Deparser: emitting header %s",
+                                         expr->toString().c_str());
+            builder->target->emitTraceMessage(builder, msgStr.c_str());
+
+            builder->emitIndent();
+            builder->appendFormat("if (%s < %s + BYTES(%s + %d)) ",
+                                  program->packetEndVar.c_str(),
+                                  program->packetStartVar.c_str(),
+                                  program->offsetVar.c_str(), width);
+            builder->blockStart();
+            builder->target->emitTraceMessage(builder,
+                                              "Deparser: invalid packet (packet too short)");
+            builder->emitIndent();
+            // We immediately return instead of jumping to reject state.
+            // It avoids reaching BPF_COMPLEXITY_LIMIT_JMP_SEQ.
+            builder->appendFormat("return %s;", builder->target->abortReturnCode().c_str());
+            builder->newline();
+            builder->blockEnd(true);
+            builder->emitIndent();
+            builder->newline();
+            unsigned alignment = 0;
+            for (auto f : headerToEmit->fields) {
+                auto ftype = deparser->program->typeMap->getType(f);
+                auto etype = EBPFTypeFactory::instance->create(ftype);
+                auto et = dynamic_cast<EBPF::IHasWidth *>(etype);
+                if (et == nullptr) {
+                    ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                            "Only headers with fixed widths supported %1%", f);
+                    return;
+                }
+                emitField(builder, f->name, expr, alignment, etype);
+                alignment += et->widthInBits();
+                alignment %= 8;
+            }
+            builder->blockEnd(true);
+        }
+    }
```

Comment:
Is doing nothing for other methods the right thing?
Should there be an `else error` here?